### PR TITLE
fix: issues with withdrawal from empty vaults and 0 shares withdrawals

### PR DESCRIPTION
**Pull request by jmonteer**
Refer to issues #247 and #246 

Added a shortcut in _shareValue to return 1:1 sharePrice if totalSupply == 0 to avoid division by zero

Checks that the user is withdrawing > 0 

Bonus: Checks that maxLoss is <= MAX_BPS

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]
Add tests for new `assert` clauses

File: tests/functional/vault/test_withdrawal.py [hunk 1]
```diff
@@ -315,3 +315,37 @@ def test_profit_degration(chain, gov, token, vault, strategy, rando):
...
+    assert token.balanceOf(gov) == 0
+
+    # Deposit something in strategies
+    chain.sleep(1)  # Needs to be a second ahead, at least
+    [s.harvest({"from": gov}) for s in strategies]
+    assert token.balanceOf(vault) < vault.totalAssets()  # Some debt is in strategies
+
+    # Trying to withdraw 0 shares. It should revert
+    with brownie.reverts():
+        vault.withdraw(0, {"from": rando})
```
Comment (fubuloubu) [inline tests/functional/vault/test_withdrawal.py:348]: This line could probably just be in another test

File: contracts/Vault.vy [hunk 2]
```diff
@@ -972,13 +976,19 @@ def withdraw(
     """
     shares: uint256 = maxShares  # May reduce this number below
 
+    # Max Loss is <=100%, revert otherwise
+    assert maxLoss <= MAX_BPS
```
Comment (fubuloubu) [inline contracts/Vault.vy:980]: missing a test case showing this

File: contracts/Vault.vy [hunk 3]
```diff
@@ -1058,10 +1068,7 @@ def pricePerShare() -> uint256:
     @dev See dev note on `withdraw`.
     @return The value of a single share.
     """
-    if self.totalSupply == 0:
-        return 10 ** self.decimals  # price of 1:1
-    else:
-        return self._shareValue(10 ** self.decimals)
+    return self._shareValue(10 ** self.decimals)
```
Comment (poolpitako) [inline contracts/Vault.vy:1071]: revert

**Review by poolpitako — APPROVED** [review]

File: tests/functional/vault/test_withdrawal.py (hunk 1 above)
Comment (jmonteer) [reply tests/functional/vault/test_withdrawal.py:348]: merging with test_progressive_withdrawal as suggested

**Comment by jmonteer** [comment]
it should be good to go @fubuloubu

File: contracts/BaseWrapper.sol [hunk 4]
```diff
@@ -179,7 +179,8 @@ abstract contract BaseWrapper {
 
                     // Limit amount to withdraw to the maximum made available to this contract
                     uint256 shares = Math.min(estimatedShares, availableShares);
-                    withdrawn = withdrawn.add(vaults[id].withdraw(shares));
+                    // If shares to be withdrawn is 0, do not try to withdraw (as it will revert)
+                    withdrawn = shares == 0 ? withdrawn : withdrawn.add(vaults[id].withdraw(shares));
```
Comment (fubuloubu) [inline contracts/BaseWrapper.sol:183]: `availableShares` is > 0 (see above guard clause), and `amount > withdrawn` should also be true (see below `break` clause)

Comment (jmonteer) [reply contracts/BaseWrapper.sol:183]: I thought the same. However for some reason tests were not passing before adding that switch...

File: contracts/Vault.vy (hunk 3 above)
Comment (fubuloubu) [reply contracts/Vault.vy:1071]: they changed `_shareValue` above too

File: tests/functional/vault/test_withdrawal.py [hunk 5]
```diff
@@ -101,6 +101,11 @@ def test_forced_withdrawal(token, gov, vault, TestStrategy, rando, chain):
 
     # Scenario 1: we panic, and try to get out as quickly as possible (total loss)
     assert token.balanceOf(rando) == 0
+
+    # User first try to withdraw with +100% losses, which are not allowed
```
Comment (fubuloubu) [inline tests/functional/vault/test_withdrawal.py:105]:
```suggestion
    # User first try to withdraw with more than 100% losses, which is nonsensical
```

File: contracts/Vault.vy [hunk 6]
```diff
@@ -1011,7 +1021,7 @@ def withdraw(
                 continue  # Nothing to withdraw from this Strategy, try the next one
 
             # Force withdraw amount from each Strategy in the order set by governance
-            loss: uint256 = Strategy(strategy).withdraw(amountNeeded)
+            loss: uint256 = Strategy(strategy).withdraw(amountNeeded)
```
Comment (fubuloubu) [inline contracts/Vault.vy:1024]: undo this change? adds extra white space

File: tests/integration/test_wrapper.py [hunk 7]
```diff
@@ -82,7 +82,6 @@ def rule_withdraw(self, ratio="st_ratio"):
         if amount > 0:
             print(f"  Wrapper.withdraw({amount})")
             self.wrapper.withdraw(amount, {"from": self.user}).return_value
-
```
Comment (fubuloubu) [inline tests/integration/test_wrapper.py:85]: undo this change, makes it easier to read

File: contracts/BaseWrapper.sol (hunk 4 above)
Comment (fubuloubu) [reply contracts/BaseWrapper.sol:183]: Can you point to this issue in a CI failure?

Comment (fubuloubu) [inline contracts/BaseWrapper.sol:183]: Prefer something more like this
```suggestion
                    // If shares to be withdrawn is 0, do not try to withdraw (as it will revert)
                    if (shares > 0) withdrawn = withdrawn.add(vaults[id].withdraw(shares));
```

Comment (jmonteer) [reply contracts/BaseWrapper.sol]: got it working without this fix.

resolving

**Comment by jmonteer** [comment]
@fubuloubu now it should be ready to go

**Review by fubuloubu — APPROVED** [review]
